Adding get started topics (#471)
osisoft/OCS-Docs: content/add-organize-data/organize-data/metadata-rules/metadata-rules-concept.md

---
uid: ccMetadataRules
---

# Metadata rules

> Whenever possible, OSIsoft recommends that metadata be explicitly included in the streams when they are being created. However, there are situations where this is not possible. In these instances you can establish a consistent naming pattern for your streams and embed metadata such as location, asset class, and asset ID in your stream names.  Then you can create metadata rules that define the pattern of the stream name. The metadata rule identifies all streams that match the defined pattern. OSIsoft Cloud Services (OCS) then parses each stream and builds out the metadata following the defined rules. 

Reviewer: Also coming from an adapter or EDS where the stream names are autogenerated but you can append or prepend information for this purpose

> Whenever possible, OSIsoft recommends that metadata be explicitly included in the streams when they are being created. However, there are situations where this is not possible. In these instances you can establish a consistent naming pattern for your streams and embed metadata such as location, asset class, and asset ID in your stream names.  Then you can create metadata rules that define the pattern of the stream name. The metadata rule identifies all streams that match the defined pattern. OSIsoft Cloud Services (OCS) then parses each stream and builds out the metadata following the defined rules. 

Reviewer 2: That is true. Maybe just `In these instances you can establish *or leverage an existing* consistent naming pattern...`. That might be a little long winded but it's more general.

> Whenever possible, OSIsoft recommends that metadata be explicitly included in the streams when they are being created. However, there are situations where this is not possible. In these instances you can establish a consistent naming pattern for your streams and embed metadata such as location, asset class, and asset ID in your stream names.  Then you can create metadata rules that define the pattern of the stream name. The metadata rule identifies all streams that match the defined pattern. OSIsoft Cloud Services (OCS) then parses each stream and builds out the metadata following the defined rules. 

Author: ```suggestion
Whenever possible, OSIsoft recommends that metadata be explicitly included in the streams when they are being created. However, there are situations where this is not possible. In these instances you can establish a consistent naming pattern for your streams and embed metadata such as location, asset class, and asset ID in your stream names.  Then you can create metadata rules that define the pattern of the stream name. The metadata rule identifies all streams that match the defined pattern. OSIsoft Cloud Services (OCS) then parses each stream and builds out the metadata following the defined rules.
```

## PI Server counterpart

Metadata rules do not have a similar counterpart in PI Server because PI points cannot store generic metadata. If a similar feature existed in PI Server, it might be a tool that runs against a list of PI points and fills in their point attributes by parsing out different parts of a structured PI point name.

## Metadata best practices  

> The following best practices are recommend to make it easier to add metadata to your streams:

Author: ```suggestion
The following best practices are recommended to make it easier to add metadata to your streams:
```

> - It is easiest to explicitly add metadata to streams and, whenever possible, OSIsoft recommends you add metadata this way.

Reviewer: not sure I understand this statement.  I know if you are creating your stream via OMF or the SDS apis I would recommend adding the metadata at stream creation time.  But after that if you are doing <10 i would do it by hand, but more than 10 I would be looking into metadata rules for automation and guaranteed consistency

Reviewer 2: This came from David Hearn, the recommendation is that it should mainly be used to leverage an existing naming pattern (or as you pointed out likely with adapters where the naming pattern can be managed but metadata cannot). If creating from an app where you can control the metadata it is preferred to do that directly. 
Maybe `When creating streams programmatically, it is easiest to explicitly add metadata to streams at creation time and, whenever possible, OSIsoft recommends you add metadata this way.`

> - It is easiest to explicitly add metadata to streams and, whenever possible, OSIsoft recommends you add metadata this way.

Author: ```suggestion
- The easiest way to explicitly add metadata is at the time that streams are created. Therefore, whenever possible, OSIsoft recommends that metadata be added during stream creation.
```

> - In situations where streams are created from an external source and you cannot explicitly include metadata fields but you are able to establish a naming pattern for stream names, adopt and apply a naming pattern that can be used with metadata rules. <!-- What are the characteristics of a naming pattern that can be used with metadata rules? -->

Author: ```suggestion
- There may be situations where streams are created from an external source and you cannot explicitly include metadata fields. If possible, establish and apply a naming pattern for stream names that can be used with metadata rules.  An example of a naming pattern is: {Region}.{Site}.{Equipment}.{Measurement}. Use delimiters to separate the parts in the naming pattern. 
```

Each part of the stream name is assigned a metadata type. The screen capture above shows the different metadata types (wildcard, metadata, and string literal) that can be assigned to each delimited part of a stream name. 

*       In this example, all data is coming from wind turbines on a wind farm, The user knows that the first part of the stream name (GEO1 in the selected stream above) identifies the wind turbine. This rule uses a wildcard to capture streams for all turbines.  

*       The second part of the stream name (P.ACT in the selected stream above) has been assigned a metadata type of metadata. Metadata is defined as key-value pairs. For example*, P.ACT* is the key for *Active Power*. The user defines all key-value pairs for this part of the stream name. 

> **Note**: Stream name parts designated as metadata are defined as key-value pairs. Each value occupying a part designated as metadata is categorized under a key defined for it. This typically takes the form of values recorded at different timestamp intervals for a particular object or device, or key. The key-value data structure facilitates analytical processing of stream data returned by a metadata rule. 

Author: I suggest removing the note.

*       In this example, all data is coming from wind turbines on a wind farm, The user knows that the first part of the stream name (GEO1 in the selected stream above) identifies the wind turbine. This rule uses a wildcard to capture streams for all turbines.  

*       The second part of the stream name (P.ACT in the selected stream above) has been assigned a metadata type of metadata. Metadata is defined as key-value pairs. For example*, P.ACT* is the key for *Active Power*. The user defines all key-value pairs for this part of the stream name. 

**Note**: Stream name parts designated as metadata are defined as key-value pairs. Each value occupying a part designated as metadata is categorized under a key defined for it. This typically takes the form of values recorded at different timestamp intervals for a particular object or device, or key. The key-value data structure facilitates analytical processing of stream data returned by a metadata rule. 

*       PV is defined as a string literal. In this example, it refers to the Pine Valley Wind Farm. 

> *       As defined, this metadata rule captures streams for all turbines for Active Power in the Pine Valley Wind Farm.

Author: ```suggestion
*       As defined, this metadata rule creates metadata for streams coming from turbines for Active Power in the Pine Valley Wind Farm.
```

## Related links

- [Create a metadata rule](xref:gpMetadataRules)

> - [Test link](xref:gsMetadataRules#step-2)

Author: ```suggestion
```
